Sources/DataConnectError.swift

```diff
@@ -14,27 +14,239 @@
 
 import Foundation
 
-/// Represents an error returned by the DataConnect service
+// MARK: Base Error Definitions
+
+/// Protocol representing an error returned by the DataConnect service
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+public protocol DataConnectError: Error, CustomDebugStringConvertible, CustomStringConvertible {
+  var message: String? { get }
+  var cause: Error? { get }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+public extension DataConnectError {
+  var debugDescription: String {
+    return "{\(Self.self), message: \(message ?? "nil"), cause: \(String(describing: cause))}"
+  }
+
+  var description: String {
+    return debugDescription
+  }
+}
+
+/// Type erased DataConnectError
 @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
-public enum DataConnectError: Error {
-  /// no firebase app specified. configure not complete
-  case appNotConfigured
+public struct AnyDataConnectError: DataConnectError {
+  private let dataConnectError: DataConnectError
+
+  init<E: DataConnectError>(dataConnectError: E) {
+    self.dataConnectError = dataConnectError
+  }
+
+  public var message: String? {
+    return dataConnectError.message
+  }
+
+  public var cause: (any Error)? {
+    return dataConnectError.cause
+  }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+/// Represents an error domain which can have more granular error codes
+public protocol DataConnectDomainError: DataConnectError {
+  associatedtype ErrorCode: DataConnectErrorCode
+
+  var code: ErrorCode { get }
+}
```

Comment:
Should this protocol be combined with `DataConnectDomainError`? Does breaking it out into its own protocol like here offer an advantage?

Reply:
I think you are referring to the DataConnectErrorCode protocol. One advantage I found was that it lets me combine the multiple protocols (StringConvertible, Equatable, Sendable, CaseIterable).

Reply:
Another reason for this was so we can have a standard type in the `var code: ErrorCode` protocol API.